<a href="https://colab.research.google.com/github/olgasem10/Unsafe/blob/main/Optuna_examples.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments
from transformers.trainer_callback import EarlyStoppingCallback
import torch
from torch.utils.data import Dataset
from sklearn.metrics import accuracy_score, precision_recall_fscore_support, roc_auc_score
import numpy as np
import pandas as pd
import optuna
from typing import Dict

### Данные (с добавлением спец токена)

In [ ]:
train_df = pd.read_csv('/kaggle/input/unsafe/train_randst0.csv')
test_df = pd.read_csv('/kaggle/input/unsafe/val_randst0.csv')
#train_df = train_df[["text", "unsafe"]]
#test_df = test_df[["text", "unsafe"]]

In [ ]:
train_df = train_df.loc[(train_df['unsafe'] >= 0.8) | (train_df['unsafe'] <= 0.2)]

In [ ]:
translator = {
    'crime_real' : 'криминал',
    'crime_web' : 'веб-криминал',
    'drugs' : 'наркотики',
    'gambling' : 'азартные игры',
    'pornography' : 'порнография',
    'prostitution' : 'проституция',
    'slavery' : 'рабство',
    'suicide' : 'суицид',
    'terrorism' : 'терроризм',
    'weapons' : 'оружие', 
    'body_shaming' : 'уродства',
    'halth_shaming' : 'инвалидность',
    'politics' : 'политика',
    'racism' : 'расизм',
    'religion' : 'религия',
    'sex_minorities' : 'гомосексуализм',
    'sexism' : 'сексизм',
    'social' : 'социальное',
    'none' : ''
}

In [ ]:
col_names = ['crime_real',
 'crime_web',
 'drugs',
 'gambling',
 'pornography',
 'prostitution',
 'slavery',
 'suicide',
 'terrorism',
 'weapons',
 'body_shaming',
 'halth_shaming',
 'politics',
 'racism',
 'religion',
 'sex_minorities',
 'sexism',
 'social']

In [ ]:
def binary(val):
    return round(val)

In [ ]:
for name in col_names:
    train_df[name] = train_df[name].apply(binary)
    test_df[name] = test_df[name].apply(binary)

In [ ]:
train_df['unsafe'] = train_df['unsafe'].apply(binary)
test_df['unsafe'] = test_df['unsafe'].apply(binary)

In [ ]:
def add_special_tokens(dataframe):
    new_dataframe = dataframe.copy()
    for column in col_names:
        indices = dataframe.index[dataframe[column] == 1]
        new_dataframe.loc[indices, ['text']] = '<' + translator[column] + '> ' + dataframe.loc[indices, ['text']].astype(str)
    return new_dataframe

In [ ]:
train_df = add_special_tokens(train_df)
test_df = add_special_tokens(test_df)

In [ ]:
#train_df = train_df.sample(frac=1)
#test_df = test_df.sample(frac=1)

In [ ]:
x_train = train_df['text'].tolist()#[:15000]
y_train = train_df['unsafe'].tolist()#[:15000]
x_test = test_df['text'].tolist()#[:2000]
y_test = test_df['unsafe'].tolist()#[:2000]

In [ ]:
len(x_test)

In [ ]:
class UnsafeData(Dataset):

    def __init__(self, texts, targets, tokenizer, max_len):
        
        super().__init__()
        
        self.texts = texts
        self.targets = targets        
        self.max_len = max_len
        self.tokenizer = tokenizer

    def __len__(self):
        
        return len(self.texts)

    
    def __getitem__(self, index):
        x = self.texts[index]

        enc_dict = self.tokenizer(x, truncation=True, max_length=self.max_len, padding='max_length')
      
        item = {key: torch.tensor(val).long() for key, val in enc_dict.items()}
        item['labels'] = torch.tensor(self.targets[index]).long()

        return item 

### Обучение

In [ ]:
def compute_metrics(pred):
    labels = pred.label_ids
    preds = pred.predictions.argmax(-1)
    precision, recall, f1, _ = precision_recall_fscore_support(labels, preds, average='weighted')
    acc = accuracy_score(labels, preds)
    return {
        'accuracy': acc,
        'f1': f1,
        'precision': precision,
        'recall': recall
    }

In [ ]:
model_name = 'DeepPavlov/rubert-base-cased-conversational'

In [ ]:
tokenizer = BertTokenizer.from_pretrained(model_name)
#model = BertForSequenceClassification.from_pretrained(model_name)

In [ ]:
train_dataset = UnsafeData(x_train, y_train, tokenizer, max_len = 60)
test_dataset = UnsafeData(x_test, y_test, tokenizer, max_len = 60)

In [ ]:
len(train_dataset), len(test_dataset)

(120120, 24501)

In [ ]:
train_dataset[10000]

{'input_ids': tensor([  101,   160, 21749,   130, 48931,   164, 40051,   477,   507,   465,
         58496,   439, 28019, 14646, 32260,  8297, 41387,   128,  1007,  1151,
         19644,   340,   362,  5703,   166,   102,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0]),
 'token_type_ids': tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
 'attention_mask': tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
 'labels': tensor(1)}

In [ ]:
def model_init():
    return BertForSequenceClassification.from_pretrained(model_name, return_dict=True)

In [ ]:
training_args = TrainingArguments(
    output_dir='/kaggle/working/bert1',
    logging_steps = 600,
    evaluation_strategy = 'steps',
    logging_dir='/kaggle/working/bert1/logs',
    save_steps = 10000,
    save_total_limit = 2,
    #metric_for_best_model  = 'f1',
    #greater_is_better = True,
    #load_best_model_at_end = True
)

trainer = Trainer(
    args=training_args,
    tokenizer=tokenizer,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    model_init=model_init,
    compute_metrics=compute_metrics,
)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
W&B installed but not logged in. Run `wandb login` or set the WANDB_API_KEY env variable.


In [ ]:
### По какой метрике оптимизировать

def my_func(metrics):
    return metrics['eval_f1']

### TPESampler

In [ ]:
### Функция для пространства парметров

def my_hp_space_optuna(trial) -> Dict[str, float]:
    return {
        "learning_rate": trial.suggest_float("learning_rate", 1e-6, 1e-4, log=True),
        "num_train_epochs": trial.suggest_int("num_train_epochs", 1, 1), #1, 2
        #"seed": trial.suggest_int("seed", 1, 40),
        "per_device_train_batch_size": trial.suggest_categorical("per_device_train_batch_size", [16, 32, 64]),
        "weight_decay": trial.suggest_float("weight_decay", 0.0, 0.1)
    }

In [ ]:
best_run = trainer.hyperparameter_search(
    hp_space = my_hp_space_optuna,
    n_trials=5,
    #n_jobs = 2,
    compute_objective = my_func,
    direction="maximize",
    backend = 'optuna',
    study_name = 'Example',
    sampler = optuna.samplers.TPESampler(),
    pruner=optuna.pruners.SuccessiveHalvingPruner() #??? #https://optuna.readthedocs.io/en/stable/reference/pruners.html#module-optuna.pruners
)

[I 2021-03-18 12:50:56,005] A new study created in memory with name: Example
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[I 2021-03-18 13:07:32,900] Trial 0 finished with value: 0.8059069203023017 and parameters: {'learning_rate': 1.4117764203025751e-05, 'num_train_epochs': 1, 'per_device_train_batch_size': 64, 'weight_decay': 0.07512942141733155}. Best is trial 0 with value: 0.8059069203023017.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[I 2021-03-18 13:39:14,688] Trial 1 finished with value: 0.7802529324828464 and parameters: {'learning_rate': 1.2779229703135447e-06, 'num_train_epochs': 1, 'per_device_train_batch_size': 16, 'weight_decay': 0.03047893234751622}. Best is trial 0 with value: 0.8059069203023017.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[I 2021-03-18 14:00:39,261] Trial 2 finished with value: 0.8050659830965247 and parameters: {'learning_rate': 4.7608224975319416e-05, 'num_train_epochs': 1, 'per_device_train_batch_size': 32, 'weight_decay': 0.04178583915566045}. Best is trial 0 with value: 0.8059069203023017.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[I 2021-03-18 14:32:20,009] Trial 3 finished with value: 0.8069978438396579 and parameters: {'learning_rate': 2.9489250021210324e-05, 'num_train_epochs': 1, 'per_device_train_batch_size': 16, 'weight_decay': 0.016346936245910295}. Best is trial 3 with value: 0.8069978438396579.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[I 2021-03-18 14:53:41,614] Trial 4 finished with value: 0.7951903321277392 and parameters: {'learning_rate': 4.240225969679652e-06, 'num_train_epochs': 1, 'per_device_train_batch_size': 32, 'weight_decay': 0.020472146538012794}. Best is trial 3 with value: 0.8069978438396579.


Результат: Best is trial 3 with value: 0.8069978438396579

parameters: {'learning_rate': 2.9489250021210324e-05, 'num_train_epochs': 1, 'per_device_train_batch_size': 16, 'weight_decay': 0.016346936245910295}

#### GridSampler

In [ ]:
search_space = {
        "learning_rate": [1e-6, 1e-4],
        "num_train_epochs": [1, 2],
        "seed": [32, 40],
        "per_device_train_batch_size": [4, 8, 16, 32, 64],
        'weight_decay': [0, 0.01, 0.001]
    }

In [ ]:
best_run = trainer.hyperparameter_search(
    n_trials=2,
    compute_objective = my_func,
    direction="maximize",
    backend = 'optuna',
    study_name = 'Example',
    sampler = optuna.samplers.GridSampler(search_space),
    pruner=optuna.pruners.SuccessiveHalvingPruner() #??? #https://optuna.readthedocs.io/en/stable/reference/pruners.html#module-optuna.pruners
)

[I 2021-03-18 15:00:10,681] A new study created in memory with name: Example
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[I 2021-03-18 15:21:35,597] Trial 0 finished with value: 0.7616865427144448 and parameters: {'learning_rate': 0.0001, 'num_train_epochs': 1, 'seed': 40, 'per_device_train_batch_size': 32}. Best is trial 0 with value: 0.7616865427144448.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased-conversational and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


[I 2021-03-18 15:54:41,419] Trial 1 finished with value: 0.7734213576011032 and parameters: {'learning_rate': 1e-06, 'num_train_epochs': 2, 'seed': 40, 'per_device_train_batch_size': 64}. Best is trial 1 with value: 0.7734213576011032.


In [ ]:
#!rm -r ./bert1

In [ ]:
trainer.save_model('rubert-base-cased-conversational-1')